scripts/controlnet.py

```diff
@@ -343,29 +374,64 @@ def build_sliders(module):
-                    base64_str = str(encoded_string, "utf-8")
-                    base64_str = "data:image/png;base64,"+ base64_str
-                    inputs['image'] = base64_str
-                return input_image.orgpreprocess(inputs)
-            return None
+       
+        def onRatioChange(h,w, percent):
+            return (int(round(self.inp_h*percent/100)), int(round(self.inp_w*percent/100)))
+
+        def set_outres(h,w):
+            return [h,w]
+        
+        def set_pref_canvas_size(h,w):
+            """return percentage to match this requested width or height"""
+            if (h == set_h_512_symbol):
+                p = 512 / self.inp_h * 100
+            if (h == set_h_768_symbol):
+                p = 768 / self.inp_h * 100
+            if (w == set_w_512_symbol):
+                p = 512 / self.inp_w * 100
+            if (w == set_w_768_symbol):
+                p = 768 / self.inp_w * 100
+            return gr.Slider.update(value=p)
+
+
+        zeroComp = gr.Number(value=0)
 
         resize_mode = gr.Radio(choices=["Envelope (Outer Fit)", "Scale to Fit (Inner Fit)", "Just Resize"], value="Scale to Fit (Inner Fit)", label="Resize Mode")
         with gr.Row():
             with gr.Column():
                 canvas_width = gr.Slider(label="Canvas Width", minimum=256, maximum=1024, value=512, step=64)
                 canvas_height = gr.Slider(label="Canvas Height", minimum=256, maximum=1024, value=512, step=64)
-                    
+
             if gradio_compat:
-                canvas_swap_res = ToolButton(value=switch_values_symbol)
+                canvas_swap_res = ToolButton(value=switch_values_symbol, label="Swap height/width")
                 canvas_swap_res.click(lambda w, h: (h, w), inputs=[canvas_width, canvas_height], outputs=[canvas_width, canvas_height])
+            
+                canvas_set_outres = ToolButton(value=set_outres_symbol, label="Set as output size")
+
+            if is_img2img:
+                canvas_set_outres.click(fn=set_outres, inputs=[canvas_height,canvas_width], outputs=[self.img2img_h_slider,self.img2img_w_slider])
+            else:
+                canvas_set_outres.click(fn=set_outres, inputs=[canvas_height,canvas_width], outputs=[self.txt2img_h_slider,self.txt2img_w_slider])
+
+            with gr.Column(variant="panel"):
+                with gr.Box():
+                    with gr.Row():
+                        canvas_ratio =  gr.Slider(label="Canvas proportional scale %", minimum=1, maximum=1000, value=100, step=1, interactive=True)
+
+                    with gr.Row(variant="panel",):
+                        canvas_sw512 = ToolButton(value=set_w_512_symbol, label="W:512")
+                        canvas_sw768 = ToolButton(value=set_w_768_symbol, label="W:768")
+                        canvas_sh512 = ToolButton(value=set_h_512_symbol, label="H:512")
+                        canvas_sh768 = ToolButton(value=set_h_768_symbol, label="H:768")
```

Comment:
> the design of the quickbuttons (512/768) could be refined somehow

They could be a bit wider.
```suggestion
                        canvas_sw512 = ToolButton(value=set_w_512_symbol, label="W:512", elem_id="btn_canvas_sw512")
                        canvas_sw768 = ToolButton(value=set_w_768_symbol, label="W:768", elem_id="btn_canvas_sw768")
                        canvas_sh512 = ToolButton(value=set_h_512_symbol, label="H:512", elem_id="btn_canvas_sh512")
                        canvas_sh768 = ToolButton(value=set_h_768_symbol, label="H:768", elem_id="btn_canvas_sh768")
```
Then add `sd-webui-controlnet/style.css`

```css
#btn_canvas_sw512, #btn_canvas_sw768, #btn_canvas_sh512, #btn_canvas_sh768 {
    max-width: 3.5em;
}
```

Reply:
thx, i add it